### PR TITLE
Migration to RestHighLevelClient

### PR DESCRIPTION
**Pull request by BiswasSouvik**
Migration from TransportClient to RestHighLevelClient.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,13 +57,9 @@ Example of agent named agent
...
   agent.sinks.es_sink.es.backoff.policy.retries=8
   agent.sinks.es_sink.es.cluster.name=es-cluster
-  agent.sinks.es_sink.es.client.transport.sniff=false
-  agent.sinks.es_sink.es.client.transport.ignore_cluster_name=false
-  agent.sinks.es_sink.es.client.transport.ping_timeout=5s
-  agent.sinks.es_sink.es.client.transport.nodes_sampler_interval=5s
   agent.sinks.es_sink.es.client.hosts=127.0.0.1:9300
-  agent.sinks.es_sink.es.index.name=defaultindex
-  agent.sinks.es_sink.es.index.type=defaulttype
+  agent.sinks.es_sink.es.index=defaultindex
```
Comment (Srinathc) [inline README.md]: what is the need to rename index.name to index?

File: README.md [hunk 2]
```diff
@@ -61,13 +57,9 @@ Example of agent named agent
...
+  agent.sinks.es_sink.es.type=defaulttype
```
Comment (Srinathc) [inline README.md]: ES Type is deprecated in ES 6.x. So we can support it internally but we can remove it from our samples.

File: flume-ng-elasticsearch6.2-sink.iml [hunk 3]
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Srinathc) [inline flume-ng-elasticsearch6.2-sink.iml]: remove the IntelliJ files.

File: src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java [hunk 4]
```diff
@@ -1,34 +1,18 @@
-/*
```
Comment (Srinathc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java]: retain the copyright

File: src/main/java/com/cognitree/flume/sink/elasticsearch/client/BulkProcessorBulider.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
+package com.cognitree.flume.sink.elasticsearch.client;
```
Comment (Srinathc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/client/BulkProcessorBulider.java]: Add copyright. The file is showing up as a new file instead of an update.

File: README.md (hunk 1 above)
Comment (BiswasSouvik) [reply README.md]: In the Constants.java, it is "es.index" instead of "es.index.name".

File: src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java [hunk 6]
```diff
@@ -58,7 +58,9 @@ public XContentBuilder serialize(Event event) {
                 logger.trace("Record in event " + data);
                 XContentParser parser = XContentFactory
                         .xContent(XContentType.JSON)
-                        .createParser(NamedXContentRegistry.EMPTY, data.toString());
+                        .createParser(NamedXContentRegistry.EMPTY,
+                                DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
```
Comment (sumitgs) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: Why THROW_UNSUPPORTED_OPERATION parameter is added here?

Comment (sumit-ct) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: Why this DeprecationHandler.THROW_UNSUPPORTED_OPERATION parameter is added here?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.cognitree.flume.sink.elasticsearch.Constants.DEFAULT_ES_PORT;
+
+/**
+ * This class creates  an instance of the {@link RestHighLevelClient}
+ * Sets the number of hosts for the client
+ */
+public class ElasticSearchClientBuilder {
```
Comment (Srinathc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java]: Rename ElasticSearchClientBuilder -> ElasticsearchClientBuilder

File: pom.xml [hunk 8]
```diff
@@ -57,13 +67,17 @@
...
-
         <dependency>
             <groupId>org.apache.flume</groupId>
             <artifactId>flume-ng-core</artifactId>
             <version>${flume-version}</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>commons-lang</groupId>
+            <artifactId>commons-lang</artifactId>
```
Comment (Srinathc) [inline pom.xml]: where do we need this?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java [hunk 9]
```diff
@@ -89,8 +81,8 @@ public Status process() throws EventDeliveryException {
                     String type = indexBuilder.getType(event);
                     String id = indexBuilder.getId(event);
                     XContentBuilder xContentBuilder = serializer.serialize(event);
-                    if(xContentBuilder != null) {
-                        if (!StringUtil.isNullOrEmpty(id)) {
+                    if (xContentBuilder != null) {
+                        if (!(StringUtils.isNotBlank(id) && StringUtils.isNotEmpty(id))) {
```
Comment (sumit-ct) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java]: What is the reason behind updating the condition for id?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java [hunk 10]
```diff
@@ -149,7 +141,7 @@ private void buildSerializer(Context context) {
             serializerClass = context.getString(ES_SERIALIZER);
         }
         this.serializer = instantiateClass(serializerClass);
-        if(this.serializer != null) {
+        if (this.serializer != null) {
```
Comment (sumit-ct) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java]: remove whitespace

File: src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    public ElasticSearchClientBuilder(String clusterName, String[] hostnames) {
+        this.clusterName = clusterName;
+        setTransportAddresses(hostnames);
+    }
+
+    public RestHighLevelClient build() {
+        RestHighLevelClient client;
+        HttpHost[] hosts = new HttpHost[transportAddresses.size()];
+        int i = 0;
+        logger.trace("Cluster Name: [{}], HostName: [{}], Port: [{}] ",
```
Comment (sumit-ct) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java]: can you add port detail in an array of arguments?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        return client;
+    }
+
+    private void setTransportAddresses(String[] transportAddresses) {
+        try {
+            this.transportAddresses = new ArrayList<>(transportAddresses.length);
+            for (String transportAddress : transportAddresses) {
+                String hostName = transportAddress.split(":")[0];
+                Integer port = transportAddress.split(":").length > 1 ?
+                        Integer.parseInt(transportAddress.split(":")[1]) : DEFAULT_ES_PORT;
```
Comment (sumit-ct) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java]: can you update this block of code to only split once and check for the array index for getting port value?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                this.transportAddresses.add(new TransportAddress(InetAddress.getByName(hostName), port));
+            }
+        } catch (Exception e) {
+            logger.error("Error in creating the TransportAddress for elasticsearch " + e.getMessage(), e);
```
Comment (sumit-ct) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/client/ElasticSearchClientBuilder.java]: Can you modify the error message? can you put specific scenario for NumberFormatException or IllegalArgumentException?

File: src/test/java/com/cognitree/flume/sink/elasticsearch/TestAvroSerializer.java [hunk 14]
```diff
@@ -100,7 +101,7 @@ public void testSerializer() throws Exception {
         XContentBuilder expected = generateContentBuilder();
         XContentBuilder actual = avroSerializer.serialize(event);
         JsonParser parser = new JsonParser();
-        assertEquals(parser.parse(expected.string()), parser.parse(actual.string()));
+        assertEquals(parser.parse(Strings.toString(expected)), parser.parse(Strings.toString(actual)));
```
Comment (sumit-ct) [inline src/test/java/com/cognitree/flume/sink/elasticsearch/TestAvroSerializer.java]: why this change?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java (hunk 6 above)
Comment (BiswasSouvik) [reply src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: It is a callback for notifying the creator of the XContentParser that parsing hit a deprecated field.

File: src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java (hunk 9 above)
Comment (BiswasSouvik) [reply src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java]: StringUtil.isNullOrEmpty() is no longer available.

File: src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java [hunk 15]
```diff
@@ -54,28 +53,21 @@
...
+            bulkProcessor = new BulkProcessorBulider().buildBulkProcessor(context, client);
         } else {
-            logger.error("Could not create transport client, No host exist");
+            logger.error("Could not create Rest client, No host exist");
         }
     }
 
     @Override
-    public Status process() throws EventDeliveryException {
+    public Sink.Status process() {
```
Comment (sumit-ct) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java]: Fix the import
